clang/include/clang/CConv/ConstraintsGraph.h

```diff
@@ -12,125 +12,269 @@
-  template <typename ConstraintType>
-  bool getNeighbors(Atom *A, std::set<Atom*> &Atoms, bool Succs) {
-    // Get the vertex descriptor.
-    auto Vidx = addVertex(A);
-    Atoms.clear();
-    if (Succs) {
-      typename boost::graph_traits<DirectedGraphType>::out_edge_iterator ei, ei_end;
-      for (boost::tie(ei, ei_end) = out_edges(Vidx, CG);
-           ei != ei_end; ++ei) {
-        auto source = boost::source(*ei, CG);
-        auto target = boost::target(*ei, CG);
-        assert(CG[source] == A && "Source has to be the given node.");
-        if (clang::dyn_cast<ConstraintType>(CG[target])) {
-          Atoms.insert(CG[target]);
-        }
-      }
-    } else {
-      typename boost::graph_traits <DirectedGraphType>::in_edge_iterator ei, ei_end;
-      for (boost::tie(ei, ei_end) = in_edges(Vidx, CG);
-           ei != ei_end; ++ei) {
-        auto source = boost::source ( *ei, CG );
-        auto target = boost::target ( *ei, CG );
-        assert(CG[target] == A && "Target has to be the given node.");
-        if (clang::dyn_cast<ConstraintType>(CG[source])) {
-          Atoms.insert(CG[source]);
-        }
-      }
-    }
-    return !Atoms.empty();
+  bool getNeighbors(Data D, std::set<Data> &DataSet, bool Succ){
+    auto *N = this->findNode(NodeType(D));
+    if (N == this->end())
+      return false;
+    DataSet.clear();
+    llvm::SetVector<EdgeType *> Edges;
+    if (Succ)
+      Edges = (*N)->getEdges();
+    else
+      Edges = (*N)->getPreds();
+    for (auto *E : Edges)
+      DataSet.insert(E->getTargetNode().getData());
+    return !DataSet.empty();
   }
 
-  vertex_t addVertex(Atom *A);
-  void removeEdge(Atom *Src, Atom *Dst);
+  bool getSuccessors(Data D, std::set<Data> &DataSet) {
+    return getNeighbors(D, DataSet, true);
+  }
+
+  bool getPredecessors(Data D, std::set<Data> &DataSet) {
+    return getNeighbors(D, DataSet, false);
+  }
+
+  void visitBreadthFirst(Data Start, llvm::function_ref<void(Data)> Fn) {
+    auto *N = this->findNode(NodeType(Start));
+    if (N == this->end())
+      return;
+    for (auto TNode : llvm::breadth_first(*N))
+      Fn(TNode->getData());
+  }
+
+protected:
+  virtual NodeType *addVertex(Data D) {
+    auto *OldN = this->findNode(NodeType(D));
+    if (OldN != this->end())
+      return *OldN;
+
+    auto *NewN = new NodeType(D);
+    this->addNode(*NewN);
+    return NewN;
+  }
 
 private:
-  std::set<ConstAtom*> AllConstAtoms;
+  template <typename G>
+  friend struct llvm::GraphTraits;
+  friend class GraphVizOutputGraph;
 };
 
-// Used during debugging to create a single graph that contains edges and nodes
-// from all constraint graphs. This single graph can then be printed to a file
-// in graphviz format.
-enum EdgeType { Checked, Ptype };
-class EdgeProperties {
+// Specialize the graph for the checked and pointer type constraint graphs. This
+// graphs stores atoms at each node.
+class ConstraintsGraph : public DataGraph<Atom *> {
```

Comment:
I'm sure this relates to my confusion about templating, above, but: How is it you can instantiate the graph with just an `Atom *` but the definition of `DataGraph` has two template/type parameters, not one?

Reply:
Ah, you have responded: The second is a default.